### ISSUE TITLE
no-unused-variable false positive when variable is declared outside of a module

### ISSUE THREAD

**Issue by adidahiya** [palantir/tslint#570]
simple repro code:

```ts
const fs = require("fs");

module Foo {
    const path = require("path");

    console.log(fs);
    console.log(path);
}
```

tslint reports `'fs'` on line 1 is unused.

**Comment by JKillian** [comment]
Possibly related to if there is a surrounding scope or not.  Problem seems to happen somewhere in the `getReferencedSymbolsForNode` function in [services.ts](https://github.com/Microsoft/TypeScript/blob/5bcb3f0be86cf251c37b618040405a654486f961/src/services/services.ts) (see below for excerpt). With a surrounding scope, the code enters the if statement below and the proper number of references are added to `result`. If there's not a surrounding scope, it enters the else branch and only the initial declaration is added to `result`. Unfortunately, unsure why at this point.

```typescript
            // Try to get the smallest valid scope that we can limit our search to;
            // otherwise we'll need to search globally (i.e. include each file).
            let scope = getSymbolScope(symbol);

            // Maps from a symbol ID to the ReferencedSymbol entry in 'result'.
            let symbolToIndex: number[] = [];
            if (scope) {
                result = [];
                getReferencesInNode(scope, symbol, declaredName, node, searchMeaning, findInStrings, findInComments, result, symbolToIndex);
            }
            else {
                let internedName = getInternedName(symbol, node, declarations)
                for (let sourceFile of sourceFiles) {
                    cancellationToken.throwIfCancellationRequested();

                    let nameTable = getNameTable(sourceFile);

                    if (lookUp(nameTable, internedName)) {
                        result = result || [];
                        getReferencesInNode(sourceFile, symbol, declaredName, node, searchMeaning, findInStrings, findInComments, result, symbolToIndex);
                    }
                }
            }
```

**Comment by JKillian** [comment]
Additional failure example:
```typescript
class HelloWorld {
    constructor(public name: string) { }
    sayHello() {
        return `Hello, ${this.name}!`;
    }
};

let hello = new HelloWorld("TSLint");
hello.sayHello();
```

This is a bad bug, going to try to get it fixed ASAP.

**Comment by myitcv** [comment]
@JKillian @adidahiya - can I ask, is this the following an instance of the same issue? Or a problem specific to `.tsx` files?

```typescript
/// <reference path="tsd.d.ts" />

import * as React from "react";
interface IMProps {
	Name: string;
}

export class HelloMsg extends React.Component<{}, {}> {
	constructor() {
		super();
	}
	render(): JSX.Element {
		let props: IMProps = { Name: "test" };

		return <div>
				<Msg {...props} />
			</div>;
	}
}

export class Msg extends React.Component<IMProps, {}> {
	constructor() {
		super();
	}
	render(): JSX.Element {
		return <div></div>;
	}
}
```


```
$ tslint -c app/tslint.json app/**/*.{ts,tsx}
app/example.tsx[13, 7]: unused variable: 'props'
```

Configuration as follows:

* TypeScript `v1.6.0-dev.20150907`
* `tslint v2.5.0-dev.5`
* `tsconfig.json`:

```
{
  "compilerOptions": {
    "target": "es5",
    "module": "system",
    "declaration": false,
    "noImplicitAny": true,
    "removeComments": false,
    "noLib": false,
    "sourceMap": true,
    "jsx": "react",
    "rootDir": ".",
    "outDir": "../dist/"
  }
}
```

* [`tslint.json`](https://gist.github.com/myitcv/783b55f3e65720de2072)

**Comment by JKillian** [comment]
@myitcv I think what you're seeing is a similar, but different problem. I'm looking into it more though

**Comment by praveennet** [comment]
I'm getting false positive when using import variables in another import statement. (due to rule 'no-unused-variable'). Below is sample code. 

```typescript
import ExternalDefinition = require("ExternalDefinition");
import Def = ExternalDefinition.MyNamespace.MyClass1;

export class MyImplementation
    extends MyBaseClasses.BaseClass1
    implements Def.Contract {

}
```

I get below error even though 'ExternalDefinition' is used in another import statement: 
> TsLint: unused variable: 'ExternalDefinition'

**Comment by myitcv** [comment]
@JKillian - just in case you haven't seen the link, I think I tracked it down to this:

https://github.com/Microsoft/TypeScript/issues/4688

**Comment by JKillian** [comment]
Thanks both. @myitcv nice find, looks like it is a TypeScript bug. Hopefully they'll get that fixed pretty soon and then we'll be sure to get that into TSLint.

@praveennet Yours is because of a TSLint bug which I discuss [here](https://github.com/Microsoft/TypeScript/issues/4560#issuecomment-139572688). I'm working on getting that fixed, (hopefully) should be released by next Friday.

**Comment by myitcv** [comment]
@JKillian - no problem. I've found that using [these tests](https://github.com/Microsoft/TypeScript/tree/master/tests/cases/fourslash) as a template for finding issues like this really useful.

**Comment by myitcv** [comment]
@JKillian @adidahiya - just so that I'm clear, are we thinking [this issue](https://github.com/palantir/tslint/issues/570#issuecomment-138332733) is a separate bug and needs fixed from TypeScript core? 

If so I'll open a new issue to track that.

**Comment by JKillian** [comment]
If the fourslash test is failing, definitely a TypeScript issue.

**Comment by hashark** [comment]
@JKillian hi,

I am using tslint version 3.14.0 and I just got a false positive similar to the one @praveennet got.
Is there a way to prevent it?

thanks!
Avichai

**Comment by JKillian** [comment]
@hashark I recommend using TS 2.0, these bugs should be fixed in it.

**Comment by hashark** [comment]
just tried and it still shows the false positive results
